File: leetcode/Recover_Binary_Search_Tree.java
import java.util.ArrayList;
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    ArrayList<TreeNode> mo = null;
    
    void midOrderTraverse(TreeNode root) {
        if (root == null)
            return;
        midOrderTraverse(root.left);
        mo.add(root);
        midOrderTraverse(root.right);
    }
    
    public void recoverTree(TreeNode root) {
        mo = new ArrayList<TreeNode>();
        midOrderTraverse(root);
        TreeNode[] tn = mo.toArray(new TreeNode[mo.size()]);
        TreeNode a = null;
        for (int i = 0; i < tn.length; ++i) {
            if ((i > 0 && tn[i].val < tn[i - 1].val) ||
                (i < tn.length - 1 && tn[i].val > tn[i + 1].val)) {
                a = tn[i];
                break;
            }
        }
        TreeNode b = null;
        for (int i = tn.length - 1; i >= 0; --i) {
            if ((i > 0 && tn[i].val < tn[i - 1].val) ||
                (i < tn.length - 1 && tn[i].val > tn[i + 1].val)) {
                b = tn[i];
                break;
            }
        }
        int t = a.val;
        a.val = b.val;
        b.val = t;
    }
}

File: leetcode/Recover_Binary_Search_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    TreeNode *rightmost_smaller_than_left;
    TreeNode *leftmost_larger_than_right;
    void findRightmostSmallerThanLeft(TreeNode *node, int &max_on_left) {
        if (node == NULL)
            return;
        findRightmostSmallerThanLeft(node->left, max_on_left);
        if (max_on_left > node->val) {
            if (rightmost_smaller_than_left == NULL ||
                rightmost_smaller_than_left->val > node->val)
                rightmost_smaller_than_left = node;
        } else {
            max_on_left = node->val;
        }
        findRightmostSmallerThanLeft(node->right, max_on_left);
    }
    void findLeftmostLargerThanRight(TreeNode *node, int &min_on_right) {
        if (node == NULL)
            return;
        findLeftmostLargerThanRight(node->right, min_on_right);
        if (min_on_right < node->val) {
            if (leftmost_larger_than_right == NULL ||
                leftmost_larger_than_right->val < node->val)
                leftmost_larger_than_right = node;
        } else {
            min_on_right = node->val;
        }
        findLeftmostLargerThanRight(node->left, min_on_right);
    }
public:
    void recoverTree(TreeNode *root) {
        rightmost_smaller_than_left = leftmost_larger_than_right = NULL;
        int max_on_left = -987654321;
        findRightmostSmallerThanLeft(root, max_on_left);
        int min_on_right = 987654321;
        findLeftmostLargerThanRight(root, min_on_right);
        
        int tmp = rightmost_smaller_than_left->val;
        rightmost_smaller_than_left->val = leftmost_larger_than_right->val;
        leftmost_larger_than_right->val = tmp;
    }
};
